### PR TITLE
[CI] Add GH action to automatically build against new JAX releases

### PR DESCRIPTION
**Pull request by dime10**
Adds an action to do the following:
> Regularly (twice a month) downloads latest JAX version, computes new commit hashes for the LLVM and MLIR-HLO dependencies, and attempts to build and test Catalyst.
> Two configurations are tested, testing the latest JAX/jaxlib package without rebuilding Catalyst with updated LLVM, and testing the latest JAX/jaxlib with rebuilding Catalyst with updated LLVM.

This workflow consists of a single job since we don't need to reuse components, thus eliminating the complexity of caching/artifact upload. The steps order is meant to test the most relevant components first (fail fast) while maintaining the order of build dependencies.

A small Python script is responsible for discovering the commit hashes of dependant repositories from a given JAX version:
> The script starts from a given jaxlib release version and attempts to collect the SHA commit hashes used by this jax release for the MLIR-HLO and LLVM dependencies.
> It uses a mix of searching specific files in the jax/tensorflow repo for hardcoded hashes and the GitHub API to translate a tensorflow monorepo hash to an equivalent commit hash in the mlir-hlo mirror repo.
>
> As of now this methods works for jaxlib v0.3.20 - v0.4.6

Also includes a few smaller fixes to the regular check Catalyst action (and other files):
- reorder jobs (also fixes job graph in the action visualization to something more reasonable)
- remove obsolete instructions
- minor formatting for uniformity
- rename changelog reminder for uniformity
- use more efficient submodule cloning:
  From my testing, `git clone <subrepo>`, `git submodule --init update <subrepo>`, and `git clone --recurse-submodules <repo>`, is very slow on GH actions compared to using the checkout action either on the main repo with the `submodules` flag, or directly on the subrepo with the `ref` flag. The reason I believe is that you cannot perform a shallow clone with the former methods. A "manual" shallow clone of the submodules is also very fast (https://stackoverflow.com/a/43136160).
- fix `.gitignore` for the *`pennylane_`*`catalyst` egg files

[sc-31521]

### CODE REVIEW

File: .github/workflows/check-catalyst.yaml [hunk 1]
```diff
@@ -129,9 +129,11 @@ jobs:
...
       if: |
         steps.cache-llvm-source.outputs.cache-hit != 'true'
-      run: |
-        git submodule update --init mlir/llvm-project
-        [ $(cd mlir/llvm-project && git rev-parse HEAD) == ${{ steps.llvm-hash.outputs.llvm-hash }} ]
+      uses: actions/checkout@v3
+      with:
+        repository: llvm/llvm-project
+        ref: ${{ steps.llvm-hash.outputs.llvm-hash }}
+        path: mlir/llvm-project
```
Comment (erick-xanadu) [inline .github/workflows/check-catalyst.yaml:136]: Could these changes also be added to the wheel scripts?

Comment (dime10) [reply .github/workflows/check-catalyst.yaml:136]: Oh good idea!

Comment (dime10) [reply .github/workflows/check-catalyst.yaml:135]: https://github.com/PennyLaneAI/catalyst/pull/45/commits/e98b40e9fd9b64428239d1d8943fa193a20b7b1b

**Review by erick-xanadu — APPROVED** [review]
:100:

**Pull request merged**